cpu/pic32/lib/pic32_rtc.c

```diff
@@ -0,0 +1,188 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * \addtogroup pic32 PIC32 Contiki Port
+ *
+ * @{
+ */
+
+/**
+ * \file   pic32_rtc.c
+ * \brief  RTC driver for PIC32MX
+ * \author Francois Berder <francois.berder@imgtec.com>
+ * \date   2017-17-02
+ */
+
+
+#ifdef __USE_RTC__
+
+#include <p32xxxx.h>
+#include <pic32_rtc.h>
+/*---------------------------------------------------------------------------*/
+static uint32_t
+bcd_to_binary(uint8_t a, uint8_t b)
+{
+  return a * 10 + b;
+}
+/*---------------------------------------------------------------------------*/
+static uint32_t
+binary_to_bcd(uint32_t a)
+{
+  uint32_t decimal = a / 10;
+  uint32_t digit = a - decimal * 10;
+
+  return (decimal << 4) | digit;
+}
+/*---------------------------------------------------------------------------*/
+int8_t
+pic32_rtc_init(void)
+{
+  SYSKEY = 0;
+  SYSKEY = 0xaa996655;
+  SYSKEY = 0x556699aa;
+
+  OSCCONSET = _OSCCON_SOSCEN_MASK;  /* Enable external 32.768kHz crystal */
+  RTCCON = _RTCCON_RTCWREN_MASK;    /* Enable write to RTCCON register */
+
+  SYSKEY = 0;
+
+  RTCALRM = 0;      /* Disable alarm */
+  RTCTIME = 0;
+  RTCDATE = 0;
+
+  /* Turn on module and disable further writes to RTCCON */
+  RTCCON = _RTCCON_ON_MASK;
+
+  /* Wait for the clock of RTC module */
+  while(!(RTCCON & _RTCCON_RTCCLKON_MASK)) {
+  }
+
+  return 0;
+}
+/*---------------------------------------------------------------------------*/
+struct date
+pic32_rtc_get_date(void)
+{
+  struct date date;
+
+  /* Make some copies to avoid rollover while reading registers */
+  uint32_t rtcdate = RTCDATE;
+  uint32_t rtctime = RTCTIME;
+
+  date.year = bcd_to_binary(
+      (rtcdate & _RTCDATE_YEAR10_MASK) >> _RTCDATE_YEAR10_POSITION,
+      (rtcdate & _RTCDATE_YEAR01_MASK) >> _RTCDATE_YEAR01_POSITION);
+
+  date.month = bcd_to_binary(
+      (rtcdate & _RTCDATE_MONTH10_MASK) >> _RTCDATE_MONTH10_POSITION,
+      (rtcdate & _RTCDATE_MONTH01_MASK) >> _RTCDATE_MONTH01_POSITION);
+
+  date.day = bcd_to_binary(
+      (rtcdate & _RTCDATE_DAY10_MASK) >> _RTCDATE_DAY10_POSITION,
+      (rtcdate & _RTCDATE_DAY01_MASK) >> _RTCDATE_DAY01_POSITION);
+
+  date.weekday = bcd_to_binary(0,
+      (rtcdate & _RTCDATE_WDAY01_MASK) >> _RTCDATE_WDAY01_POSITION);
+
+  date.hour = bcd_to_binary(
+      (rtctime & _RTCTIME_HR10_MASK) >> _RTCTIME_HR10_POSITION,
+      (rtctime & _RTCTIME_HR01_MASK) >> _RTCTIME_HR01_POSITION);
+
+  date.minute = bcd_to_binary(
+      (rtctime & _RTCTIME_MIN10_MASK) >> _RTCTIME_MIN10_POSITION,
+      (rtctime & _RTCTIME_MIN01_MASK) >> _RTCTIME_MIN01_POSITION);
+
+  date.second = bcd_to_binary(
+      (rtctime & _RTCTIME_SEC10_MASK) >> _RTCTIME_SEC10_POSITION,
+      (rtctime & _RTCTIME_SEC01_MASK) >> _RTCTIME_SEC01_POSITION);
+
+  return date;
+}
+/*---------------------------------------------------------------------------*/
+int8_t
+pic32_rtc_set_date(struct date date)
+{
+  uint32_t rtccon_state = RTCCON;
+
+  if(date.year > 99
```

Comment:
should we separate conditions of date and time, this if condition is way too big